File: ontrack-extension-stash/src/main/java/net/nemerosa/ontrack/extension/stash/property/StashProjectConfigurationPropertyType.java
package net.nemerosa.ontrack.extension.stash.property;

import com.fasterxml.jackson.databind.JsonNode;
import net.nemerosa.ontrack.common.MapBuilder;
import net.nemerosa.ontrack.extension.git.property.AbstractGitProjectConfigurationPropertyType;
import net.nemerosa.ontrack.extension.stash.StashExtensionFeature;
import net.nemerosa.ontrack.extension.stash.model.StashConfiguration;
import net.nemerosa.ontrack.extension.stash.service.StashConfigurationService;
import net.nemerosa.ontrack.json.JsonUtils;
import net.nemerosa.ontrack.model.security.ProjectConfig;
import net.nemerosa.ontrack.model.security.SecurityService;
import net.nemerosa.ontrack.model.structure.ProjectEntity;
import net.nemerosa.ontrack.model.structure.ProjectEntityType;
import net.nemerosa.ontrack.model.support.ConfigurationPropertyType;
import org.jetbrains.annotations.NotNull;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import java.util.EnumSet;
import java.util.Set;
import java.util.function.Function;

@Component
public class StashProjectConfigurationPropertyType extends AbstractGitProjectConfigurationPropertyType<StashProjectConfigurationProperty>
        implements ConfigurationPropertyType<StashConfiguration, StashProjectConfigurationProperty> {

    private final StashConfigurationService configurationService;

    @Autowired
    public StashProjectConfigurationPropertyType(StashExtensionFeature extensionFeature, StashConfigurationService configurationService) {
        super(extensionFeature);
        this.configurationService = configurationService;
    }

    @Override
    public String getName() {
        return "Bitbucket Server configuration";
    }

    @Override
    public String getDescription() {
        return "Associates the project with a Bitbucket Server repository";
    }

    @Override
    public Set<ProjectEntityType> getSupportedEntityTypes() {
        return EnumSet.of(ProjectEntityType.PROJECT);
    }

    @Override
    public boolean canEdit(ProjectEntity entity, SecurityService securityService) {
        return securityService.isProjectFunctionGranted(entity.projectId(), ProjectConfig.class);
    }

    @Override
    public boolean canView(ProjectEntity entity, SecurityService securityService) {
        return true;
    }

    @Override
    public StashProjectConfigurationProperty fromClient(JsonNode node) {
        return fromStorage(node);
    }

    @Override
    public StashProjectConfigurationProperty fromStorage(JsonNode node) {
        String configurationName = node.path("configuration").asText();
        // Looks the configuration up
        StashConfiguration configuration = configurationService.getConfiguration(configurationName);
        // OK
        return new StashProjectConfigurationProperty(
                configuration,
                node.path("project").asText(),
                node.path("repository").asText(),
                node.path("indexationInterval").asInt(),
                JsonUtils.get(node, "issueServiceConfigurationIdentifier", null)
        );
    }

    @Override
    public JsonNode forStorage(StashProjectConfigurationProperty value) {
        return format(
                MapBuilder.params()
                        .with("configuration", value.getConfiguration().getName())
                        .with("project", value.getProject())
                        .with("repository", value.getRepository())
                        .with("indexationInterval", value.getIndexationInterval())
                        .with("issueServiceConfigurationIdentifier", value.getIssueServiceConfigurationIdentifier())
                        .get()
        );
    }

    @Override
    public StashProjectConfigurationProperty replaceValue(@NotNull StashProjectConfigurationProperty value, Function<String, String> replacementFunction) {
        return new StashProjectConfigurationProperty(
                value.getConfiguration(),
                replacementFunction.apply(value.getProject()),
                replacementFunction.apply(value.getRepository()),
                value.getIndexationInterval(),
                value.getIssueServiceConfigurationIdentifier()
        );
    }

}

File: ontrack-extension-stash/src/main/java/net/nemerosa/ontrack/extension/stash/property/StashProjectConfigurationPropertyType.kt
package net.nemerosa.ontrack.extension.stash.property

import com.fasterxml.jackson.databind.JsonNode
import net.nemerosa.ontrack.extension.git.property.AbstractGitProjectConfigurationPropertyType
import net.nemerosa.ontrack.extension.stash.StashExtensionFeature
import net.nemerosa.ontrack.extension.stash.model.StashConfiguration
import net.nemerosa.ontrack.extension.stash.service.StashConfigurationService
import net.nemerosa.ontrack.json.asJson
import net.nemerosa.ontrack.json.getTextField
import net.nemerosa.ontrack.model.security.ProjectConfig
import net.nemerosa.ontrack.model.security.SecurityService
import net.nemerosa.ontrack.model.structure.ProjectEntity
import net.nemerosa.ontrack.model.structure.ProjectEntityType
import net.nemerosa.ontrack.model.support.ConfigurationPropertyType
import org.springframework.stereotype.Component
import java.util.*
import java.util.function.Function

@Component
class StashProjectConfigurationPropertyType(
    extensionFeature: StashExtensionFeature,
    private val configurationService: StashConfigurationService
) : AbstractGitProjectConfigurationPropertyType<StashProjectConfigurationProperty>(extensionFeature),
    ConfigurationPropertyType<StashConfiguration, StashProjectConfigurationProperty> {
    override val name: String = "Bitbucket Server configuration"

    override val description: String = "Associates the project with a Bitbucket Server repository"

    override val supportedEntityTypes: Set<ProjectEntityType> = EnumSet.of<ProjectEntityType>(ProjectEntityType.PROJECT)

    override fun canEdit(entity: ProjectEntity, securityService: SecurityService): Boolean {
        return securityService.isProjectFunctionGranted(entity.projectId(), ProjectConfig::class.java)
    }

    override fun canView(entity: ProjectEntity, securityService: SecurityService): Boolean {
        return true
    }

    override fun fromClient(node: JsonNode): StashProjectConfigurationProperty {
        return fromStorage(node)
    }

    override fun fromStorage(node: JsonNode): StashProjectConfigurationProperty {
        val configurationName = node.path("configuration").asText()
        // Looks the configuration up
        val configuration = configurationService.getConfiguration(configurationName)
        // OK
        return StashProjectConfigurationProperty(
            configuration = configuration,
            project = node.path("project").asText(),
            repository = node.path("repository").asText(),
            indexationInterval = node.path("indexationInterval").asInt(),
            issueServiceConfigurationIdentifier = node.getTextField("issueServiceConfigurationIdentifier")
        )
    }

    override fun forStorage(value: StashProjectConfigurationProperty): JsonNode {
        return mapOf(
            "configuration" to value.configuration.name,
            "project" to value.project,
            "repository" to value.repository,
            "indexationInterval" to value.indexationInterval,
            "issueServiceConfigurationIdentifier" to value.issueServiceConfigurationIdentifier,
        ).asJson()
    }

    @Deprecated("Will be removed in V5")
    override fun replaceValue(
        value: StashProjectConfigurationProperty,
        replacementFunction: Function<String, String>
    ): StashProjectConfigurationProperty {
        return StashProjectConfigurationProperty(
            value.configuration,
            replacementFunction.apply(value.project),
            replacementFunction.apply(value.repository),
            value.indexationInterval,
            value.issueServiceConfigurationIdentifier
        )
    }
}
